lucene/core/src/java/org/apache/lucene/util/fst/FST.java

```diff
@@ -632,36 +641,85 @@ long addNode(Builder<T> builder, Builder.UnCompiledNode<T> nodeIn) throws IOExce
     if (doFixedArray) {
       final int MAX_HEADER_SIZE = 11; // header(byte) + numArcs(vint) + numBytes(vint)
       assert maxBytesPerArc > 0;
-      // 2nd pass just "expands" all arcs to take up a fixed
-      // byte size
+      // 2nd pass just "expands" all arcs to take up a fixed byte size
+
+      // If more than 1/2 of the "slots" would be occupied, write an arc array that may have holes in it
+      // so that we can address the arcs directly by label without binary search
+      // TODO: There's no need to write the labels in this case; we can just store the first label in the header
+      // TODO: tune this heuristic?
+      // TODO: write multiple direct blocks with gaps between?
+      int labelRange = nodeIn.arcs[nodeIn.numArcs - 1].label - nodeIn.arcs[0].label + 1;
+      boolean writeDirectly = builder.useDirectArcAddressing && labelRange > 0 && labelRange < 4 * nodeIn.numArcs;
 
       //System.out.println("write int @pos=" + (fixedArrayStart-4) + " numArcs=" + nodeIn.numArcs);
       // create the header
       // TODO: clean this up: or just rewind+reuse and deal with it
       byte header[] = new byte[MAX_HEADER_SIZE]; 
       ByteArrayDataOutput bad = new ByteArrayDataOutput(header);
       // write a "false" first arc:
-      bad.writeByte(ARCS_AS_FIXED_ARRAY);
-      bad.writeVInt(nodeIn.numArcs);
+      if (writeDirectly) {
+        bad.writeByte(ARCS_AS_DIRECT_ARRAY);
```

Comment:
Perhaps change constants to something reflecting reality here -- ARCS_AS_ARRAY_DENSE, ARCS_AS_ARRAY_WITH_GAPS?